integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-spark-service-bundle/nifi-spark-service/src/main/java/com/thinkbiganalytics/nifi/v2/spark/SparkJobserverService.java

```diff
@@ -0,0 +1,353 @@
+                params.put(ISparkJobServerClientConstants.PARAM_SYNC, "true");
+                params.put(ISparkJobServerClientConstants.PARAM_TIMEOUT, syncTimeout);
+            }
+
+            jobResult = client.startJob(args, params);
+
+            getLogger().info(String.format("Executed %s %s on context %s on Spark Jobserver %s", appName, classPath, contextName, jobServerUrl));
+
+        } catch (Exception ex) {
+            getLogger().error(ex.getMessage());
+        }
+        finally {
+            synchronized (sparkContextsStates) {
+                if (sparkContextsStates.containsKey(contextName)) {
+                    sparkContextsStates.get(contextName).resetTimeoutTime();
+                    sparkContextsStates.get(contextName).removeExecutionLock(id);
+                }
+            }
+        }
+
+        if (jobResult != null && jobResult.getStatus() != "ERROR") {
+            success = true;
+        }
+        return success;
+    }
+
+    /**
+     * Using a timer task to periodically delete Spark contexts which have timed out
+     */
+    private void setupContextTimeoutTask() {
+
+        getLogger().info("Creating context timeout task");
+        TimerTask contextTimeoutTask = new TimerTask() {
+            @Override
+            public void run() {
+                synchronized (sparkContextsStates) {
+                    if (!sparkContextsStates.isEmpty()) {
+                        for (SparkContextState sparkContextState: sparkContextsStates.values()) {
+                            if (!sparkContextState.isLocked() && sparkContextState.hasTimedOut()) {
+                                getLogger().info(String.format("Found context %s which timed out", sparkContextState.ContextName));
+                                deleteContext(sparkContextState.ContextName);
+                            }
+                        }
+                    }
+                }
+            }
+        };
+        long delay = 0;
+        long intervalPeriod = 10000;
+
+        // schedules the task to be run in an interval
+        contextTimeoutTimer.scheduleAtFixedRate(contextTimeoutTask, delay,
+                                                intervalPeriod);
+        getLogger().info("Created and scheduled context timeout task");
+    }
+
+    /**
+     * Stop timer task and kill all running Spark contexts.
+     */
+    @OnShutdown
+    public void shutdown() {
+        contextTimeoutTimer.cancel();
+        for (SparkContextState sparkContextState: sparkContextsStates.values()) {
+            deleteContext(sparkContextState.ContextName);
+        }
+    }
+
+    /**
+     * Checks to see if a Spark Context is new on the Controller Service
+     * @param contextName the name of the Context
+     * @return true or false
+     */
+    private boolean isNewContext(String contextName) {
+        boolean newContext = false;
+        synchronized (sparkContextsStates) {
+            try {
+                while (true) {
```

Comment:
This block could be refactored for readability. Please use a flag to indicate whether it should loop rather than `true`.